# 04 EDA — Order Items

Read-only inspection of `olist_order_items_dataset.csv`. No cleaning, charts, or feature engineering.

## 1. Import libraries

Import pandas for tabular inspection.

In [1]:
import pandas as pd

**Observation:** pandas is ready.

## 2. Load CSV

Load only this CSV into memory. Timestamp columns are converted on the in-memory copy.

In [2]:
file_path = '../data/raw/olist/olist_order_items_dataset.csv'
df = pd.read_csv(file_path)
for column in ['shipping_limit_date']:
    df[column] = pd.to_datetime(df[column], errors='coerce')

**Observation:** The table is loaded; the raw CSV is unchanged.

## 3. Shape

Check table size.

In [3]:
df.shape

(112650, 7)

**Observation:** This reports the table's row and column count.

## 4. First 5 rows

Preview the raw structure.

In [4]:
df.head()

                           order_id  order_item_id  ...   price freight_value
0  00010242fe8c5a6d1ba2dd792cb16214              1  ...   58.90         13.29
1  00018f77f2f0320c557190d7a144bdd3              1  ...  239.90         19.93
2  000229ec398224ef6ca0657da4fc703e              1  ...  199.00         17.87
3  00024acbcdf0a6daa1e931b038114c75              1  ...   12.99         12.79
4  00042b26cf59d7ce69dfabb4e55b4fd9              1  ...  199.90         18.14

[5 rows x 7 columns]

**Observation:** The preview confirms the row grain and field layout.

## 5. Column names

List available fields.

In [5]:
df.columns.tolist()

['order_id', 'order_item_id', 'product_id', 'seller_id', 'shipping_limit_date', 'price', 'freight_value']

**Observation:** These are the fields available for this table's analysis.

## 6. Data types

Review pandas data types.

In [6]:
df.dtypes

order_id                          str
order_item_id                   int64
product_id                        str
seller_id                         str
shipping_limit_date    datetime64[us]
price                         float64
freight_value                 float64
dtype: object

**Observation:** Types show which fields are text, numeric, or converted datetime values.

## 7. Missing values

Count nulls by field.

In [7]:
df.isna().sum()

order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64

**Observation:** Missing counts identify incomplete fields without changing them.

## 8. Missing percentage

Measure missingness relative to table size.

In [8]:
(df.isna().mean().mul(100).round(2)).sort_values(ascending=False)

order_id               0.0
order_item_id          0.0
product_id             0.0
seller_id              0.0
shipping_limit_date    0.0
price                  0.0
freight_value          0.0
dtype: float64

**Observation:** Percentages show the relative scale of incomplete fields.

## 9. Duplicate rows

Count complete row duplicates.

In [9]:
df.duplicated().sum()

np.int64(0)

**Observation:** This is the number of repeated full rows; none are removed.

## 10. Duplicate primary keys

Validate the documented candidate key.

In [10]:
primary_key = ['order_id', 'order_item_id']
df.duplicated(subset=primary_key, keep=False).sum()

np.int64(0)

**Observation:** This counts rows involved in duplicate candidate keys for order_id, order_item_id.

## 11. Unique values

Inspect meaningful category or entity counts.

In [11]:
pd.Series({'orders_with_items': df['order_id'].nunique(), 'products': df['product_id'].nunique(), 'sellers': df['seller_id'].nunique(), 'largest_item_sequence': df['order_item_id'].max()})

orders_with_items        98666
products                 32951
sellers                   3095
largest_item_sequence       21
dtype: int64

**Observation:** This gives a concise view of important distinct values.

## 12. Basic describe()

Generate a basic summary for all fields.

In [12]:
df.describe(include='all')

                                order_id  ...  freight_value
count                             112650  ...  112650.000000
unique                             98666  ...            NaN
top     8272b63d03f5f79c56e9e4120aec44ef  ...            NaN
freq                                  21  ...            NaN
mean                                 NaN  ...      19.990320
min                                  NaN  ...       0.000000
25%                                  NaN  ...      13.080000
50%                                  NaN  ...      16.260000
75%                                  NaN  ...      21.150000
max                                  NaN  ...     409.680000
std                                  NaN  ...      15.806405

[11 rows x 7 columns]

**Observation:** The summary describes completeness, distinctness, common values, and numeric ranges where applicable.

## 13. Table-specific observations

Order items bridge orders to products and sellers. The table contains 112,650 item rows across 98,666 orders.

## 14. Data quality observations

No missing or exact duplicate rows were found. The composite key (`order_id`, `order_item_id`) is valid, and all order, product, and seller references have matching parents.